jsv2/kv/kv.go

```diff
@@ -0,0 +1,434 @@
+// Note that if no option is specified, the default is 30 minutes. You can set
+// this option to a negative value to instruct to always remove the markers,
+// regardless of their age.
+type DeleteMarkersOlderThan time.Duration
+
+func (ttl DeleteMarkersOlderThan) configurePurge(opts *purgeOpts) error {
+	opts.dmthr = time.Duration(ttl)
+	return nil
+}
+
+type DeleteOpt interface {
+	configureDelete(opts *deleteOpts) error
+}
+
+type deleteOpts struct {
+	// Remove all previous revisions.
+	purge bool
+
+	// Delete only if the latest revision matches.
+	revision uint64
+}
+
+type deleteOptFn func(opts *deleteOpts) error
+
+func (opt deleteOptFn) configureDelete(opts *deleteOpts) error {
+	return opt(opts)
+}
+
+// LastRevision deletes if the latest revision matches.
+func LastRevision(revision uint64) DeleteOpt {
+	return deleteOptFn(func(opts *deleteOpts) error {
+		opts.revision = revision
+		return nil
+	})
+}
+
+// purge removes all previous revisions.
+func purge() DeleteOpt {
+	return deleteOptFn(func(opts *deleteOpts) error {
+		opts.purge = true
+		return nil
+	})
+}
+
+// KeyValueConfig is for configuring a KeyValue store.
+type KeyValueConfig struct {
+	Bucket       string
+	Description  string
+	MaxValueSize int32
+	History      uint8
+	TTL          time.Duration
+	MaxBytes     int64
+	Storage      jetstream.StorageType
+	Replicas     int
+	Placement    *jetstream.Placement
+}
+
+// Used to watch all keys.
+const (
+	KeyValueMaxHistory = 64
+	AllKeys            = ">"
+	kvLatestRevision   = 0
+	kvop               = "KV-Operation"
+	kvdel              = "DEL"
+	kvpurge            = "PURGE"
+)
+
+type KeyValueOp uint8
+
+const (
+	KeyValuePut KeyValueOp = iota
+	KeyValueDelete
+	KeyValuePurge
+)
+
+func (op KeyValueOp) String() string {
+	switch op {
+	case KeyValuePut:
+		return "KeyValuePutOp"
+	case KeyValueDelete:
+		return "KeyValueDeleteOp"
+	case KeyValuePurge:
+		return "KeyValuePurgeOp"
+	default:
+		return "Unknown Operation"
+	}
+}
+
+// KeyValueEntry is a retrieved entry for Get or List or Watch.
+type KeyValueEntry interface {
+	// Bucket is the bucket the data was loaded from.
+	Bucket() string
+	// Key is the key that was retrieved.
+	Key() string
+	// Value is the retrieved value.
+	Value() []byte
+	// Revision is a unique sequence for this value.
+	Revision() uint64
+	// Created is the time the data was put in the bucket.
+	Created() time.Time
+	// Delta is distance from the latest value.
+	Delta() uint64
+	// Operation returns Put or Delete or Purge.
+	Operation() KeyValueOp
+}
+
+// Errors
+var (
+	ErrKeyValueConfigRequired = errors.New("nats: config required")
+	ErrInvalidBucketName      = errors.New("nats: invalid bucket name")
+	ErrInvalidKey             = errors.New("nats: invalid key")
+	ErrBucketNotFound         = errors.New("nats: bucket not found")
+	ErrBadBucket              = errors.New("nats: bucket not valid key-value store")
+	ErrKeyNotFound            = errors.New("nats: key not found")
```

Comment:
This ErrKeyNotFound might benefit from some error wrapping too, maybe have something like a `KeyValueError` struct and mention that `errors.Is` is preferred for some of the new checks.

Reply:
I decided to add proper KV support in a separate PR, to be able to make it at least a little bit more granular.